### PR TITLE
Error when requesting a gpu higher than the available ones

### PR DESCRIPTION
**Pull request by emcastillo**
This closes:#7309
Merge after #7552 
The main problem is that when using a device_id of '@cupy:id' a GPU object is created in chainer associated to the cupy device creation. However, neither chainer or cupy verify that the device_id is less than the number of available gpus when first creating the device handler object.

ChainerX does this check, but chainerx device handles are only created if the device name is other than numpy, cupy or intel64 (chainer/backend.py:157)

### CODE REVIEW

**Comment by niboshi** [comment]
Related: #7470

**Comment by emcastillo** [comment]
@niboshi I refactored the code as suggested. Still working on the tests and
checking the `MultiprocessParallelUpdater` issue with creating the cuda context before.

Thanks!

**Comment by emcastillo** [comment]
I will remove the fork code and update the review accordingly once I find the reason behind the pickle error

**Comment by emcastillo** [comment]
This should be ready for review now

File: chainer/backends/cuda.py [hunk 1]
```diff
@@ -207,6 +207,11 @@ class GpuDevice(_backend.Device):
     def __init__(self, device):
         check_cuda_available()
         assert isinstance(device, Device)
+        num_gpus = cupy.cuda.runtime.getDeviceCount()
+        if (device.id < 0) or (device.id >= num_gpus):
```
Comment (niboshi) [inline chainer/backends/cuda.py]: You can write `not 0 <= device.id < num_gpus` which is easier to read.

File: chainer/backends/cuda.py [hunk 2]
```diff
@@ -207,6 +207,11 @@ class GpuDevice(_backend.Device):
...
+            raise ValueError(
+                'Invalid CUDA device ID must be lower '
```
Comment (niboshi) [inline chainer/backends/cuda.py]: "Invalid" should be removed.

File: chainer/backends/cuda.py [hunk 3]
```diff
@@ -207,6 +207,11 @@ class GpuDevice(_backend.Device):
...
+                'than max gpus:{} id: {} '.format(num_gpus, device.id))
```
Comment (niboshi) [inline chainer/backends/cuda.py]: Extra space at the end.

File: chainer/backends/cuda.py [hunk 4]
```diff
@@ -271,7 +276,18 @@ def get_device_from_id(device_id):
             returns.
     """
     if device_id is not None:
+        # There are two code paths,
```
Comment (niboshi) [inline chainer/backends/cuda.py]: Is this comment relevant to the code?

File: chainer/backends/cuda.py [hunk 5]
```diff
@@ -348,6 +364,12 @@ def _get_cuda_device(*args):
     for arg in args:
         if type(arg) is not bool and isinstance(arg, _integer_types):
             check_cuda_available()
+            # Arg is an integer with the id of the device for cupy
+            num_gpus = cupy.cuda.runtime.getDeviceCount()
+            if (arg < 0) or (arg >= num_gpus):
```
Comment (niboshi) [inline chainer/backends/cuda.py]: ditto.

File: chainer/backends/cuda.py [hunk 6]
```diff
@@ -348,6 +364,12 @@ def _get_cuda_device(*args):
...
+                raise ValueError(
+                    'Invalid CUDA device ID must be in the range '
+                    'of max gpus:{} id: {} '.format(num_gpus, arg))
```
Comment (niboshi) [inline chainer/backends/cuda.py]: ditto.

File: tests/chainer_tests/backends_tests/test_cuda.py [hunk 7]
```diff
@@ -108,6 +108,11 @@ def test_get_device_warning(self):
     def test_get_device_from_id(self):
         assert cuda.get_device_from_id(0) == cuda.Device(0)
 
+    @attr.gpu
+    def test_get_device_from_id_invalid(self):
+        with self.assertRaises(ValueError):
+            cuda.get_device_from_id(20)
```
Comment (niboshi) [inline tests/chainer_tests/backends_tests/test_cuda.py]: 20-GPU machine is not that unrealistic. How about something like 99999?

Comment (niboshi) [inline tests/chainer_tests/backends_tests/test_cuda.py]: Test with -1 as well.

File: tests/chainer_tests/backends_tests/test_cuda.py [hunk 8]
```diff
@@ -119,6 +124,13 @@ def test_get_device_for_int(self):
             device = cuda.get_device(0)
         assert device == cuda.Device(0)
 
+    @attr.gpu
+    def test_get_device_for_int_invalid(self):
+        with self.assertRaises(ValueError):
+            with testing.assert_warns(DeprecationWarning):
+                device = cuda.get_device(20)
+            assert device == cuda.Device(20)
```
Comment (niboshi) [inline tests/chainer_tests/backends_tests/test_cuda.py]: Will this line ever executed?

**Comment by emcastillo** [comment]
Updated

File: chainer/backends/cuda.py [hunk 9]
```diff
@@ -207,6 +207,11 @@ class GpuDevice(_backend.Device):
     def __init__(self, device):
         check_cuda_available()
         assert isinstance(device, Device)
+        num_gpus = cupy.cuda.runtime.getDeviceCount()
+        if not 0 <= device.id < num_gpus:
+            raise ValueError(
+                'CUDA device ID must be lower '
+                'than max gpus:{} id: {}'.format(num_gpus, device.id))
```
Comment (niboshi) [inline chainer/backends/cuda.py]: I think this representation is difficult to interpret.
How about something like this:
"CUDA device ID must be lower than max GPUs. (actual: {deivce.id} > max: {num_gpus})"

**Comment by emcastillo** [comment]
PTAL
